File: src/ru/hudspeed/pro/RadarScanGate.java
package ru.gpsantiradar.app;

/** Separates radar-scan eligibility from committing a successfully completed scan. */
public final class RadarScanGate {
    private static final long GPS_RECOVERY_MILLIS = 5000L;
    private long lastObservedFixElapsed;
    private boolean recoveryPending;

    public Decision assess(long elapsedRealtime, boolean hasAccuracy,
                           float accuracyMeters, double movementMeters) {
        float requiredMovement = hasAccuracy ? Math.max(3f, accuracyMeters) : 3f;
        if (lastObservedFixElapsed > 0L
                && elapsedRealtime - lastObservedFixElapsed > GPS_RECOVERY_MILLIS) {
            recoveryPending = true;
        }
        lastObservedFixElapsed = elapsedRealtime;
        boolean scanRequested = requiredMovement < 60f
                && movementMeters > requiredMovement;
        return new Decision(elapsedRealtime, scanRequested,
                scanRequested && recoveryPending);
    }

    public void accept(Decision decision) {
        if (decision == null || !decision.scanRequested) {
            throw new IllegalArgumentException("Only a requested radar scan can be accepted");
        }
        recoveryPending = false;
    }

    public static final class Decision {
        final long elapsedRealtime;
        public final boolean scanRequested;
        public final boolean gpsRecovered;

        Decision(long elapsedRealtime, boolean scanRequested, boolean gpsRecovered) {
            this.elapsedRealtime = elapsedRealtime;
            this.scanRequested = scanRequested;
            this.gpsRecovered = gpsRecovered;
        }
    }
}

File: src/main/kotlin/ru/gpsantiradar/app/RadarScanGate.kt
package ru.gpsantiradar.app

import kotlin.math.max

/** Separates radar-scan eligibility from committing a successfully completed scan. */
class RadarScanGate {
    private var lastObservedFixElapsed = 0L
    private var recoveryPending = false

    fun assess(elapsedRealtime: Long, hasAccuracy: Boolean, accuracyMeters: Float, movementMeters: Double): Decision {
        val requiredMovement = if (hasAccuracy) max(3f, accuracyMeters) else 3f
        if (lastObservedFixElapsed > 0 && elapsedRealtime - lastObservedFixElapsed > GPS_RECOVERY_MILLIS) recoveryPending = true
        lastObservedFixElapsed = elapsedRealtime
        val scanRequested = requiredMovement < 60f && movementMeters > requiredMovement
        return Decision(elapsedRealtime, scanRequested, scanRequested && recoveryPending)
    }

    fun accept(decision: Decision?) {
        require(decision != null && decision.scanRequested) { "Only a requested radar scan can be accepted" }
        recoveryPending = false
    }

    class Decision internal constructor(
        val elapsedRealtime: Long,
        val scanRequested: Boolean,
        val gpsRecovered: Boolean
    )

    companion object { private const val GPS_RECOVERY_MILLIS = 5000L }
}
